sdk/python/kfp/compiler/_container_builder.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import tarfile
+import tempfile
+import os
+import uuid
+
+class ContainerBuilder(object):
+  """
+  ContainerBuilder helps build a container image
+  """
+  def __init__(self, gcs_staging):
+    if not self._check_gcs_path(gcs_staging):
+      raise Exception('ContainerBuilder __init__ failure.')
+    self._gcs_staging = gcs_staging
+
+  def _check_gcs_path(self, gcs_path):
+    """ _check_gcs_path check both the path validity and write permissions """
+    logging.info('Checking path: {}...'.format(gcs_path))
+    if not gcs_path.startswith('gs://'):
+      logging.error('Error: {} should be a GCS path.'.format(gcs_path))
+      return False
+    return True
+
+  def _generate_kaniko_spec(self, namespace, context, docker_filename, target_image):
+    """_generate_kaniko_yaml generates kaniko job yaml based on a template yaml """
+    content = {
+        'apiVersion': 'v1',
+        'metadata': {
+            'generateName': 'kaniko-',
+            'namespace': namespace,
+        },
+        'kind': 'Pod',
+        'spec': {
+            'restartPolicy': 'Never',
+            'containers': [{
+                'name': 'kaniko',
+                'args': ['--cache=true',
+                         '--dockerfile=' + docker_filename,
+                         '--context=' + context,
+                         '--destination=' + target_image],
+                'image': 'gcr.io/kaniko-project/executor@sha256:78d44ec4e9cb5545d7f85c1924695c89503ded86a59f92c7ae658afa3cff5400',
+                'env': [{
+                    'name': 'GOOGLE_APPLICATION_CREDENTIALS',
```

Comment:
Does kaniko support docker registry other than GCR? If yes, we may consider to make the pod secret to be configurable.

Reply:
+1 But keep in mind that this is a refactoring. The changes should come next.
Optimally the refactoring PRs should be separate from feature changes.

Reply:
Agree. Please open bug to track the discussions and feel free to close the comment here.

Reply:
Bug created: https://github.com/kubeflow/pipelines/issues/1700